src/coreclr/jit/emit.cpp

```diff
@@ -4102,6 +4122,315 @@ void emitter::emitDispCommentForHandle(size_t handle, GenTreeFlags flag)
+
+                    if (checkedPreviousInstruction && instructionDescriptor->idIns() == jmp->idIns() &&
+                        jmp == instructionDescriptor)
+                    {
+                        CLANG_FORMAT_COMMENT_ANCHOR
+                        // the last instruction in the group is the jmp we're looking for
+                        // and it jumps to the next instruction group so we don't really need it
+
+                        if (!previousInstructionWasCall)
+                        {
+#ifdef DEBUG
+                            if (EMITVERBOSE)
+                            {
+                                printf("Removing unconditional jump [%08X/%03u] from the last instruction in "
+                                       "IG%02u to the next IG IG%02u label %s\n",
+                                       dspPtr(jmp), jmp->idDebugOnlyInfo()->idNum, group->igNum, targetGroup->igNum,
+                                       emitLabelString(targetGroup));
+                            }
+#endif
+                            UNATIVE_OFFSET sizeRemoved = instructionDescriptor->idCodeSize();
+
+                            // set state needed at the end of the loop in ADJUST_GROUP
+                            adjIG      = sizeRemoved;
+                            lstIG      = group;
+                            jmpRemoved = true;
+
+                            // unlink the jump
+                            if (previousJmp != nullptr)
+                            {
+                                previousJmp->idjNext = jmp->idjNext;
+                                if (jmp == emitJumpLast)
+                                {
+                                    emitJumpLast = previousJmp;
+                                }
+                            }
+                            else
+                            {
+                                assert(jmp == emitJumpList);
+                                emitJumpList = jmp->idjNext;
+                            }
+                            jmp->idjNext = nullptr;
+
+#if DEBUG
+                            // clear the instruction data
+                            memset((BYTE*)instructionDescriptor, 0, instructionDescriptor->idCodeSize());
+#endif
+
+                            // remove the instruction from the group
+                            group->igInsCnt -= 1;
+                            group->igSize -= (unsigned short)sizeRemoved;
+                            emitTotalCodeSize -= sizeRemoved;
+
+                            // flag the group as resized and recounted
+                            group->igFlags |= (IGF_UPD_ISZ | IGF_UPD_ICOUNT);
+
+                            // cleanup
+                            instructionDescriptor = nullptr;
+                            jmp                   = nullptr;
+                            removedCount += 1;
+
+                            goto ADJUST_GROUP;
```

Comment:
I am not sure if I follow the logic of having `goto ADJUST_GROUP`. Why can't be done here? Also, if I reading the code correctly, you eliminate 1 jump at a time and again iterate from the 1st jump? Why is that?

Imagine having 20 jumps with 18th, 19th and 20th unconditional jumps that can be eliminated. From what I understand, you will remove the 18st jump, then again iterate 17 jumps until you reach 19th jump?

Reply:
This is because the logic was original part of, and then copied from, `emitJumpDistBind` which does this. And yes it is wasteful. This is the part that I've mentioned I can probably linearize to a single pass.

Reply:
I would prefer seeing that refactoring in this PR rather than a follow up PR.

Reply:
Sure. If the general concensus is that this PR is worth continuing to work on that's fine. 
As I said I just didn't want to sink more time in if the overall cost benefit ratio was already too bad to consider adding this.

Reply:
It is definitely worth continuing :)

Reply:
This has been updated. there is now a single pass over the jmp list.